### PR TITLE
Add corosync and pacemaker packages

### PR DESCRIPTION
**Pull request by AokiNobuhiko-Metatechno**
# Purpose of pull request

This pr adds following packges.
* corosync package based on meta-openembedded's meta-networking/recipes-extended/corosync/corosync_3.0.3.bb
* pacemaker package based on meta-cgl's meta-cgl-common/recipes-cgl/pacemaker/pacemaker_2.0.3.bb

In addition, adds following packges.

These are corosync's dependencies.
* libsctp
* libqb
* libknet

These are pacemaker's dependencies.
* openhpi
* libopenipmi
* cluster-glue
* resource-agents

# Test
## How to test

Add following line to conf/local.conf.
```
IMAGE_INSTALL_append = " corosync pacemaker"
```

The following was confirmed for corosync.
* `which` command
* Simple Execution

For packages with dependencies, the existence of files was confirmed.

Run following commands from console.
1. `corosync`
```
# which corosync
# systemctl start pacemaker
# systemctl status corosync
```
2. `corosync-blackbox`
```
# which corosync-blackbox
# corosync-blackbox
```
3. `corosync-xmlproc`
```
# which corosync-xmlproc
# cat <<EOS > test_corosync-xmlproc.xml
<?xml version="1.0"?>

<corosync>

<totem version="2" crypto_cipher="none" crypto_hash="none">
    <interface ringnumber="0" bindnetaddr="192.168.1.0" mcastaddr="239.255.1.1"
        mcastport="5405" ttl="1" />
</totem>

</corosync>
EOS
# corosync-xmlproc test_corosync-xmlproc.xml
```
4. `corosync-cfgtool`
```
# which corosync-cfgtool
# corosync-cfgtool -s
```
5. `corosync-cmapctl`
```
# which corosync-cmapctl
# corosync-cmapctl
```
6. `corosync-cpgtool`
```
# which corosync-cpgtool
# corosync-cpgtool
# echo $?
```
7. `corosync-keygen`
```
# which corosync-keygen
# corosync-keygen
# ls -la /etc/corosync/authkey
```
8. `corosync-quorumtool`
```
# which corosync-quorumtool
# corosync-quorumtool
```
9. `pacemakerd`
```
# which pacemakerd
# systemctl start pacemaker
# systemctl status pacemaker
```
10. `crm_attribute`
```
# which crm_attribute
# crm_attribute --node localhost --name location --update office
# crm_attribute --node localhost --name location --query
```
11. `crm_node`
```
# which crm_node
# crm_node -l
```
12. `fence_legacy`
```
# which fence_legacy
# fence_legacy -h
```

## Test result

1. `corosync`
```
root@qemuarm64:~# which corosync
/usr/sbin/corosync
root@qemuarm64:~# systemctl start pacemaker
root@qemuarm64:~# systemctl status corosync | cat
● corosync.service - Corosync Cluster Engine
   Loaded: loaded (/lib/systemd/system/corosync.service; disabled; vendor preset: enabled)
   Active: active (running) since Thu 2023-03-02 06:14:49 UTC; 13s ago
     Docs: man:corosync
           man:corosync.conf
           man:corosync_overview
 Main PID: 181 (corosync)
    Tasks: 9 (limit: 517)
   Memory: 179.5M
   CGroup: /system.slice/corosync.service
           └─181 /usr/sbin/corosync -f

Mar 02 06:14:47 qemuarm64 systemd[1]: Starting Corosync Cluster Engine...
Mar 02 06:14:49 qemuarm64 systemd[1]: Started Corosync Cluster Engine.
```
2. `corosync-blackbox`
```
root@qemuarm64:~# which corosync-blackbox
/usr/bin/corosync-blackbox
root@qemuarm64:~# corosync-blackbox
[debug] shm size:8392704; real_size:8392704; rb->word_size:2098176
[debug] read total of: 8392724
Dumping the contents of /var/lib/corosync/fdata
...<snip>...
```
3. `corosync-xmlproc`
```
root@qemuarm64:~# which corosync-xmlproc
/usr/bin/corosync-xmlproc
root@qemuarm64:~# corosync-xmlproc test_corosync-xmlproc.xml
# Corosync configuration file generated from test_corosync-xmlproc.xml at 2023-03-02T06:16:06Z

totem {
    version: 2
    crypto_cipher: none
    crypto_hash: none

    interface {
        ringnumber: 0
        bindnetaddr: 192.168.1.0
        mcastaddr: 239.255.1.1
        mcastport: 5405
        ttl: 1
    }
}
```
4. `corosync-cfgtool`
```
root@qemuarm64:~# which corosync-cfgtool
/usr/sbin/corosync-cfgtool
root@qemuarm64:~# corosync-cfgtool -s
Printing link status.
Local node ID 1
LINK ID 0
        addr    = 127.0.0.1
        status:
                node 0: link enabled:1  link connected:1
```
5. `corosync-cmapctl`
```
root@qemuarm64:~# which corosync-cmapctl
/usr/sbin/corosync-cmapctl
root@qemuarm64:~# corosync-cmapctl
config.totemconfig_reload_in_progress (u8) = 0
internal_configuration.service.0.name (str) = corosync_cmap
...<snip>...
```
6. `corosync-cpgtool`
```
root@qemuarm64:~# which corosync-cpgtool
/usr/sbin/corosync-cpgtool
root@qemuarm64:~# corosync-cpgtool
Group Name             PID         Node ID
crmd\x00
                       198               1 (127.0.0.1)
attrd\x00
                       196               1 (127.0.0.1)
stonith-ng\x00
                       194               1 (127.0.0.1)
cib\x00
                       193               1 (127.0.0.1)
pacemakerd\x00
                       192               1 (127.0.0.1)
root@qemuarm64:~# echo $?
0
```
7. `corosync-keygen`
```
root@qemuarm64:~# which corosync-keygen
/usr/sbin/corosync-keygen
root@qemuarm64:~# corosync-keygen
Corosync Cluster Engine Authentication key generator.
Gathering 2048 bits for key from /dev/urandom.
Writing corosync key to /etc/corosync/authkey.
root@qemuarm64:~# ls -la /etc/corosync/authkey
-r--------    1 root     root           256 Mar  2 06:17 /etc/corosync/authkey
```
8. `corosync-quorumtool`
```
root@qemuarm64:~# which corosync-quorumtool
/usr/sbin/corosync-quorumtool
root@qemuarm64:~# corosync-quorumtool
Quorum information
------------------
Date:             Thu Mar  2 06:17:58 2023
Quorum provider:  corosync_votequorum
Nodes:            1
Node ID:          1
Ring ID:          1/4
Quorate:          Yes

Votequorum information
----------------------
Expected votes:   1
Highest expected: 1
Total votes:      1
Quorum:           1  
Flags:            Quorate 

Membership information
----------------------
    Nodeid      Votes Name
         1          1 node1 (local)
root@qemuarm64:~# 
```

9. `pacemakerd`
```
root@qemuarm64:~# which pacemakerd
/usr/sbin/pacemakerd
root@qemuarm64:~# systemctl status pacemaker | cat
● pacemaker.service - Pacemaker High Availability Cluster Manager
   Loaded: loaded (/lib/systemd/system/pacemaker.service; disabled; vendor preset: enabled)
   Active: active (running) since Thu 2023-03-02 06:14:49 UTC; 3min 38s ago
     Docs: man:pacemakerd
           https://clusterlabs.org/pacemaker/doc/en-US/Pacemaker/2.0/html-single/Pacemaker_Explained/index.html
 Main PID: 192 (pacemakerd)
    Tasks: 7
   Memory: 44.8M
   CGroup: /system.slice/pacemaker.service
           ├─192 /usr/sbin/pacemakerd -f
           ├─193 /usr/lib/pacemaker/pacemaker-based
           ├─194 /usr/lib/pacemaker/pacemaker-fenced
           ├─195 /usr/lib/pacemaker/pacemaker-execd
           ├─196 /usr/lib/pacemaker/pacemaker-attrd
           ├─197 /usr/lib/pacemaker/pacemaker-schedulerd
           └─198 /usr/lib/pacemaker/pacemaker-controld

Mar 02 06:14:49 qemuarm64 systemd[1]: Started Pacemaker High Availability Cluster Manager.
```
10. `crm_attribute`
```
root@qemuarm64:~# which crm_attribute
/usr/sbin/crm_attribute
root@qemuarm64:~# crm_attribute --node localhost --name location --update office
root@qemuarm64:~# crm_attribute --node localhost --name location --query
scope=nodes  name=location value=office
```
11. `crm_node`
```
root@qemuarm64:~# which crm_node
/usr/sbin/crm_node
root@qemuarm64:~# crm_node -l
1 node1 member
```
12. `fence_legacy`
```
root@qemuarm64:~# which fence_legacy
/usr/sbin/fence_legacy
root@qemuarm64:~# fence_legacy -h
Helper that presents a Pacemaker-style interface for Linux-HA stonith plugins

Should never be invoked by the user directly


Usage: fence_legacy [options]

Options:
  -h               usage
  -t <sub agent>   sub agent
  -n <name>        nodename
  -o <string>      Action:  on | off | reset (default) | stat | hostlist
  -s <stonith>     stonith command
  -q               quiet mode
  -V               version
```

### CODE REVIEW

File: recipes-debian/cluster-glue/cluster-glue_debian.bb [hunk 1]
```diff
@@ -0,0 +1,176 @@
+#base recipe: meta-cgl/meta-cgl-common/recipes-cgl/cluster-glue/cluster-glue_1.0.12.bb
+#base branch: dunfell
+#base commit: b8529657086ba4d308ee8cc99b720aaedaa3ae18
+
+DESCRIPTION = "Cluster Glue is a set of libraries, tools and utilities suitable for \
+the Heartbeat/Pacemaker cluster stack. In essence, Glue is everything that \
+is not the cluster messaging layer (Heartbeat), nor the cluster resource manager \
+(Pacemaker), nor a Resource Agent."
+HOMEPAGE = "http://clusterlabs.org/"
+LICENSE = "GPLv2 & LGPLv2.1"
```
Comment (masami256) [inline recipes-debian/cluster-glue/cluster-glue_debian.bb:10]: There is two files under the BSD -3-clause but I'm not sure these files are used or not.

```
Files: replace/daemon.c
       replace/alphasort.c
Copyright: 1998-2015 The Regents of the University of California
License: BSD-3-clause
```

https://metadata.ftp-master.debian.org/changelogs//main/c/cluster-glue/cluster-glue_1.0.12-12_copyright

File: recipes-debian/libsctp/libsctp_debian.bb [hunk 2]
```diff
@@ -0,0 +1,64 @@
+#base recipe: meta-networking/recipes-support/lksctp-tools/lksctp-tools_1.0.18.bb
+#base branch: dunfell
+#base commit: 6792ebdd966aa0fb662989529193a0940fbfee00
+
+SUMMARY = "The Linux Kernel Stream Control Transmission Protocol (lksctp) project"
+HOMEPAGE = "http://lksctp.org"
+SECTION = "net"
+LICENSE = "LGPLv2.1 & GPLv2"
```
Comment (masami256) [inline recipes-debian/libsctp/libsctp_debian.bb]: according to the debian copyright file, there is some BSD-3-clause license files.

```
Files: src/withsctp/*
Copyright: 2003 La Monte HP Yarroll <piggy@acm.org>
License: BSD-3-clause
```

File: recipes-debian/corosync/corosync/corosync [hunk 3]
```diff
@@ -0,0 +1,116 @@
+#! /bin/sh
```
Comment (masami256) [inline recipes-debian/corosync/corosync/corosync:1]: This file is not installed on sysv init environment.

```
root@qemuarm64:~# ls /etc/init.d/
banner.sh                 dbus-1                    hostname.sh               ofono                     reboot                    stop-bootlogd             urandom
bluetooth                 devpts.sh                 hwclock.sh                pacemaker                 rmnologin.sh              sysfs.sh
bootlogd                  dmesg.sh                  modutils.sh               populate-volatile.sh      run-postinsts             syslog
bootmisc.sh               dropbear                  mountall.sh               rc                        save-rtc.sh               systemd-udevd
checkroot.sh              functions                 mountnfs.sh               rcS                       sendsigs                  umountfs
connman                   halt                      networking                read-only-rootfs-hook.sh  single                    umountnfs.sh
```

File: recipes-debian/libsctp/libsctp_debian.bb (hunk 2 above)
Comment (AokiNobuhiko-Metatechno) [reply recipes-debian/libsctp/libsctp_debian.bb]: The log.do_compile contains the log of building `withsctp/`.
Therefore, add the BSD-3-clause license.
```
Making all in withsctp
make[3]: Entering directory '<snip>/libsctp/1.0.18-r0/lksctp-tools-1.0.18/src/withsctp'
```

File: recipes-debian/cluster-glue/cluster-glue_debian.bb (hunk 1 above)
Comment (AokiNobuhiko-Metatechno) [reply recipes-debian/cluster-glue/cluster-glue_debian.bb:10]: The log.do_compile does not contain the log of building the two files.

File: recipes-debian/corosync/corosync/corosync (hunk 3 above)
Comment (AokiNobuhiko-Metatechno) [reply recipes-debian/corosync/corosync/corosync:1]: I am sorry. I fixed it.
```
root@qemuarm64:/etc/init.d# ./corosync start
Starting corosync daemon: corosyncMar 03 04:18:31 notice  [MAIN  ] Corosync Cluster Engine 3.0.1 starting up
Mar 03 04:18:31 info    [MAIN  ] Corosync built-in features: dbus systemd xmlconf snmp relro bindnow
.
root@qemuarm64:/etc/init.d# ./corosync status
corosync is running.
root@qemuarm64:/etc/init.d# ./corosync stop
Stopping corosync daemon: corosync.
root@qemuarm64:/etc/init.d# ./pacemaker start
Starting Pacemaker Cluster Manager[  OK  ]
root@qemuarm64:/etc/init.d# ./pacemaker status
pacemakerd (pid 2149) is running...
root@qemuarm64:/etc/init.d# ./pacemaker stop
Signaling Pacemaker Cluster Manager to terminate[  OK  ]
Waiting for cluster services to unload[  OK  ]
```

File: recipes-debian/cluster-glue/cluster-glue_debian.bb (hunk 1 above)
Comment (masami256) [reply recipes-debian/cluster-glue/cluster-glue_debian.bb:10]: I see. Thank you for checking it.

File: recipes-debian/libsctp/libsctp_debian.bb (hunk 2 above)
Comment (masami256) [reply recipes-debian/libsctp/libsctp_debian.bb]: Thanks.

File: recipes-debian/corosync/corosync/corosync (hunk 3 above)
Comment (masami256) [reply recipes-debian/corosync/corosync/corosync:1]: thanks.

**Comment by masami256** [comment]
Tested on both systemd and sysvinit environments that worked fine.

**Pull request merged**